### PR TITLE
Add RowNumber operator

### PR DESCRIPTION
**Pull request by mbasmanova**

### CODE REVIEW

File: velox/exec/RowNumber.cpp [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  }
+  return *result->as<FlatVector<int64_t>>();
+}
+
+RowVectorPtr RowNumber::getOutput() {
+  if (input_ == nullptr) {
+    return nullptr;
+  }
+
+  if (table_) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:132]: nit: handle single partition case first?
```
if (table_ == nullptr) {
  // No partition keys.
  return getOutputForSinglePartition();
}

...
````

File: velox/exec/RowNumber.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+  // No partition keys.
+  return getOutputForSinglePartition();
+}
+
+RowVectorPtr RowNumber::getOutputForSinglePartition() {
+  const auto numInput = input_->size();
+
+  vector_size_t numOutput;
+  if (limit_) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:186]:
```
add a check?
VELOX_LE(numTotalInput_, limit_.value());
```

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova looks great. Thanks!

**Comment by mbasmanova** [comment]
@xiaoxmeng Meng, thank you for review. Updated the PR to address comments.

File: velox/exec/tests/RowNumberTest.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+TEST_F(RowNumberTest, noPartitionKeys) {
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>(1'000, [](auto row) { return row; }),
+  });
+
+  createDuckDbTable({data, data});
+
+  // No limit.
+  auto plan = PlanBuilder().values({data, data}).rowNumber({}).planNode();
```
Comment (aditi-pandit) [inline velox/exec/tests/RowNumberTest.cpp:59]: The plan and testlimit lambdas are repeated in each test function. Maybe abstract them to a common function in the parent class.

**Comment by mbasmanova** [comment]
@aditi-pandit Aditi, thank you for review. Updated the PR.

**Review by kewang1024 — COMMENTED** [review]
Two noob question:
1. How do we track the memory of the operator of RowNumber?
2. I was referencing the logic of Java version and saw that we need outputChannels in java, but here we don't need, is it because for velox, the data fed to RowNumber is already pruned whereas for java version, it's not?

**Comment by mbasmanova** [comment]
@kewang1024 

> How do we track the memory of the operator of RowNumber?

Each operator has a MemoryPool that's used for all significant memory allocations. MemoryPool allows to track the total amount of memory being used.

> I was referencing the logic of Java version and saw that we need outputChannels in java, but here we don't need, is it because for velox, the data fed to RowNumber is already pruned whereas for java version, it's not?

Not sure I understand the question. Any chance you could elaborate a bit?

**Comment by aditi-pandit** [comment]
@mbasmanova, @xiaoxmeng  : Curious, will this operator have spill to disk ?

**Pull request closed without merging**